kernel/hConstr.ml

```diff
@@ -0,0 +1,378 @@
+  | Construct (c,u) -> combinesmall 11 (combine (Construct.SyntacticOrd.hash c) (UVars.Instance.hash u))
+  | Case (_,u,pms,(p,_),_,c,bl) ->
+    let hash_ctx (bnd,c) =
+      Array.fold_left (fun hash na -> combine (hash_annot na) hash) c.hash bnd
+    in
+    let hpms = hash_array hash pms in
+    let hbl = hash_array hash_ctx bl in
+    let h = combine5 (UVars.Instance.hash u)
+        c.hash hpms (hash_ctx p) hbl
+    in
+    combinesmall 12 h
+  | Fix (_,(lna,tl,bl)) ->
+    combinesmall 13 (combine3 (hash_array hash_annot lna) (hash_array hash tl) (hash_array hash bl))
+  | CoFix (_,(lna,tl,bl)) ->
+    combinesmall 14 (combine3 (hash_array hash_annot lna) (hash_array hash tl) (hash_array hash bl))
+  | Meta _ -> assert false
+  | Rel n -> combinesmall 16 n
+  | Proj (p,_,c) -> combinesmall 17 (combine (Projection.SyntacticOrd.hash p) c.hash)
+  | Int i -> combinesmall 18 (Uint63.hash i)
+  | Float f -> combinesmall 19 (Float64.hash f)
+  | String s -> combinesmall 20 (Pstring.hash s)
+  | Array (u,t,def,ty) -> combinesmall 21 (combine4 (UVars.Instance.hash u) (hash_array hash t) def.hash ty.hash)
+
+let kind_to_constr = function
+  | Rel n -> mkRel n
+  | Var i -> mkVar i
+  | Meta _ | Evar _ -> assert false
+  | Sort s -> mkSort s
+  | Cast (c,k,t) -> mkCast (c.self,k,t.self)
+  | Prod (na,t,c) -> mkProd (na,t.self,c.self)
+  | Lambda (na,t,c) -> mkLambda (na,t.self,c.self)
+  | LetIn (na,b,t,c) -> mkLetIn (na,b.self,t.self,c.self)
+  | App (h,args) -> mkApp (h.self, Array.map self args)
+  | Const cu -> mkConstU cu
+  | Ind iu -> mkIndU iu
+  | Construct cu -> mkConstructU cu
+  | Case (ci,u,pms,(p,r),iv,c,bl) ->
+    let to_ctx (bnd,c) = bnd, c.self in
+    let iv = match iv with
+      | NoInvert -> NoInvert
+      | CaseInvert x -> CaseInvert {indices=Array.map self x.indices}
+    in
+    mkCase (ci,u,Array.map self pms,(to_ctx p,r),iv,c.self,Array.map to_ctx bl)
+  | Fix (ln,(lna,tl,bl)) ->
+    mkFix (ln,(lna,Array.map self tl,Array.map self bl))
+  | CoFix (ln,(lna,tl,bl)) ->
+    mkCoFix (ln,(lna,Array.map self tl,Array.map self bl))
+  | Proj (p,r,c) -> mkProj (p,r,c.self)
+  | Int i -> mkInt i
+  | Float f -> mkFloat f
+  | String s -> mkString s
+  | Array (u,t,def,ty) -> mkArray (u,Array.map self t,def.self,ty.self)
+
+let hcons_inplace f a = Array.iteri (fun i x -> Array.unsafe_set a i (f x)) a
+
+let of_kind_nohashcons = function
+  | App (c, [||]) -> c
+  | kind ->
+  let self = kind_to_constr kind in
+  let hash = hash_kind kind in
+  let () = match kind with
+    | Rel _ -> assert false
+    | _ -> ()
+  in
+  { self; kind; hash; isRel = 0; refcount = 1 }
+
+let rec of_constr tbl local_env c =
+  let c =
+    let kind = of_constr_aux tbl local_env c in
+    let self = kind_to_constr kind in
+    let self = if hasheq_kind (Constr.kind self) (Constr.kind c) then c else self in
+    let hash = hash_kind kind in
+    let isRel, hash = match kind with
+      | Rel n ->
+        let uid = Range.get local_env.rels (n-1) in
```

Comment:
TODO do something on unbound rels (say isRel = 0 ?)

Reply:
That should be an anomaly, if anything. Why do you want to handle this?

Reply:
We already handle it in typeops, but the current code of this PR would mask it. Why should we stop handling it?
Didn't we want to move towards having less anomalies in the kernel?

Reply:
Why not simply raise an Type_errors.UnboundRel then?

Reply:
we don't have an env

Reply:
Worst case is that you just raise here an exception defined in the current mli and catch it in Typeops.

Reply:
and what do we do in typeops? we still won't have the correct env

Reply:
What? In execute we have the environment the term is supposed to live in, where is the issue?

Reply:
execute needs a hconstr, what are we calling it on if HConstr.of_constr raises?